internal/server/triple.go

```diff
@@ -90,54 +111,125 @@ func (s *Server) GetTriples(ctx context.Context, in *pb.GetTriplesRequest) (
 
 	// Observation DCIDs.
 	if len(obsDcids) > 0 {
-		for _, param := range []struct {
-			predKey, pred string
-		}{
-			{obsAncestorTypeObservedNode, "observedNode"},
-			{obsAncestorTypeComparedNode, "comparedNode"},
-		} {
-			rowList := buildObservedNodeKey(obsDcids, param.predKey)
-			baseDataMap, branchDataMap, err := bigTableReadRowsParallel(
-				ctx, s.store, rowList,
-				func(dcid string, raw []byte) (interface{}, error) {
-					return string(raw), nil
-				}, nil)
-			if err != nil {
-				return nil, err
-			}
-			// Map from observation dcid to observedNode dcid.
-			observedNodeMap := map[string]string{}
+		if svobsMode {
+			dcidList := ""
 			for _, dcid := range obsDcids {
-				if data, ok := branchDataMap[dcid]; ok {
-					observedNodeMap[dcid] = data.(string)
-				} else if data, ok := baseDataMap[dcid]; ok {
-					observedNodeMap[dcid] = data.(string)
-				}
+				dcidList += fmt.Sprintf("\"%s\" ", dcid)
 			}
-			// Get the observedNode names.
-			observedNodes := []string{}
-			for _, dcid := range observedNodeMap {
-				observedNodes = append(observedNodes, dcid)
+			selectStatment := "SELECT ?o ?provenance "
+			tripleStatment := "?o typeOf StatVarObservation . ?o provenance ?provenance . "
+			for _, prop := range obsProps {
+				selectStatment += fmt.Sprintf("?%s ", prop.name)
+				tripleStatment += fmt.Sprintf("?o %s ?%s . ", prop.name, prop.name)
 			}
-			nameRowList := buildPropertyValuesKey(observedNodes, "name", true)
-			nameNodes, err := readPropertyValues(ctx, s.store, nameRowList)
+			tripleStatment += fmt.Sprintf("?o dcid (%s)", dcidList)
+			sparql := fmt.Sprintf(
+				`%s
+				WHERE {
+					%s
+				}
+				`, selectStatment, tripleStatment,
+			)
+			resp, err := s.Query(ctx, &pb.QueryRequest{Sparql: sparql})
 			if err != nil {
 				return nil, err
 			}
-			for dcid, observedNode := range observedNodeMap {
-				if _, exist := resultsMap[dcid]; !exist {
-					resultsMap[dcid] = []*Triple{}
+			for _, row := range resp.GetRows() {
+				dcid := row.GetCells()[0].Value
+				prov := row.GetCells()[1].Value
+				objDcids := []string{}
+				objTriples := map[string]*Triple{}
+				for i, prop := range obsProps {
+					if row.GetCells()[i+2].Value != "" {
+						if prop.isObj {
+							// The object is a node; need to fetch the name.
+							objDcid := row.GetCells()[i+2].Value
```

Comment:
nit: use a variable for row.GetCells()[i+2].Value ?

Reply:
Done